### ISSUE TITLE
Support custom instrumentation for generic gRPC servers

### ISSUE THREAD

**Issue by be9** [timostamm/protobuf-ts#201]
@timostamm re: this feature I would like to discuss possible implementation approaches before I get into coding.

Our gRPC servers (that we plan to migrate to `protobuf-ts` generic servers) heavily use two things:

* all methods are instrumented (latency is measured and exposed via Prometheus metrics)
* methods extract tracing context from gRPC metadata and do something with it.

Here's one of existing wrappers for native `grpc` callbacks: [gist](https://gist.github.com/be9/b0cd255ec76937030e8c639cfc39b165).
I want to package this nicely in the new server library that I create using `protobuf-ts`, but `adaptService` cannot inject any additional code to callbacks that it generates and `ServerCallContext` interface is not extendable.

Here's what I have in mind:

* Add callbacks (like, `beforeUnaryCall`, `afterUnaryCall`, etc.) in a form of options object passed to `adaptService`. These callbacks would receive `ServerCallContext` as an argument.
* Additional fields like tracing context can be set by the callback. Method bodies would get an extended interface as argument:
  ```ts
  interface RicherContext extends ServerCallContext {
    tracingContext: TracingContext;
  }

  class CoolService implements ICoolService {
    // ...
    async meth(request: SomeRequest, context: RicherContext): Promise<SomeResponse> {
      // ...
    }
  }
  ```

Are there better implementation options here, what do you think?

**Comment by timostamm** [comment]
I think there are multiple steps here:

**1) allow user-provided context in generated interface** 
For the context, the main issue is that we generate a service interface that expects you to use `ServerCallContext`:
https://github.com/timostamm/protobuf-ts/blob/e4db65f7e6bed6309852b1b68d0456959ba99bb3/packages/example-node-grpc-backend-server/service-example.server.ts#L12-L16

I think we can safely change the code generator to generate:

```ts
export interface IExampleService<T = ServerCallContext> {
    /**
     * @generated from protobuf rpc: Unary(spec.ExampleRequest) returns (spec.ExampleResponse);
     */
    unary(request: ExampleRequest, context: T): Promise<ExampleResponse>;
```

The default typing means that it would still work exactly as before, but it could be easily customized by the user. 

This is something protobuf-ts can safely support without regrets :) 

**2) accept user-provided context in `adaptService()` of `@protobuf-ts/grpc-backend`**

To use user-provided contexts, `adaptService()` will need to change. It will have to accept a new optional argument `contextFactory`. It probably needs to have the same signature as [`createContext`](https://github.com/timostamm/protobuf-ts/blob/e4db65f7e6bed6309852b1b68d0456959ba99bb3/packages/grpc-backend/src/grpc-adapter.ts#L78) - although this is probably a good chance to update this signature, and pass `ServiceInfo` as well, not just `MethodInfo`. The `contextFactory` will obviously need to be passed down to where `createContext()` is called right now, and replace it if set.

I think this is a very reasonable change to make full use of 1).

**3) Server side callbacks**

For the callbacks, I'm not sure if this will turn out perfect on the first try. I think there are also use cases for full-fledged server-side interceptors that can do asynchronous operations. Once it's released as part of protobuf-ts, it's not easy to revise, so I'm a bit wary here. 

In order to get a bit more confidence, would it be an option to wrap the service? You could do that either before or after passing it to `adaptService()`. You can walk through the `ServiceInfo` and use `localName` to get the method names.

**Comment by be9** [comment]
@timostamm thanks for feedback, Timo! I've just opened #204 that implements your points 1 and 2. 

Wrapping will be good enough, so let's cross this out

**Comment by timostamm** [comment]
This is released in [v2.2.2](https://github.com/timostamm/protobuf-ts/releases/tag/v2.2.2)
